score.go

```diff
@@ -0,0 +1,885 @@
+	// This is the number of invalid messages in the topic.
+	// The value of the parameter is a counter, decaying with InvalidMessageDeliveriesDecay.
+	// The weight of the parameter MUST be negative.
+	InvalidMessageDeliveriesWeight, InvalidMessageDeliveriesDecay float64
+}
+
+type peerStats struct {
+	// true if the peer is currently connected
+	connected bool
+
+	// expiration time of the score stats for disconnected peers
+	expire time.Time
+
+	// per topc stats
+	topics map[string]*topicStats
+
+	// IP tracking; store as string for easy processing
+	ips []string
+}
+
+type topicStats struct {
+	// true if the peer is in the mesh
+	inMesh bool
+
+	// time when the peer was (last) GRAFTed; valid only when in mesh
+	graftTime time.Time
+
+	// time in mesh (updated during refresh/decay to avoid calling gettimeofday on
+	// every score invocation)
+	meshTime time.Duration
+
+	// first message deliveries
+	firstMessageDeliveries float64
+
+	// mesh message deliveries
+	meshMessageDeliveries float64
+
+	// true if the peer has been enough time in the mesh to activate mess message deliveries
+	meshMessageDeliveriesActive bool
+
+	// sticky mesh rate failure penalty counter
+	meshFailurePenalty float64
+
+	// invalid message counter
+	invalidMessageDeliveries float64
+}
+
+type peerScore struct {
+	sync.Mutex
+
+	// the score parameters
+	params *PeerScoreParams
+
+	// per peer stats for score calculation
+	peerStats map[peer.ID]*peerStats
+
+	// IP colocation tracking
+	peerIPs map[string]map[peer.ID]struct{}
+
+	// message delivery tracking
+	deliveries *messageDeliveries
+
+	msgID MsgIdFunction
+	host  host.Host
+}
+
+type messageDeliveries struct {
+	records map[string]*deliveryRecord
+
+	// queue for cleaning up old delivery records
+	head *deliveryEntry
+	tail *deliveryEntry
+}
+
+type deliveryRecord struct {
+	status    int
+	validated time.Time
+	peers     map[peer.ID]struct{}
+}
+
+type deliveryEntry struct {
+	id     string
+	expire time.Time
+	next   *deliveryEntry
+}
+
+// delivery record status
+const (
+	delivery_unknown   = iota // we don't know (yet) if the message is valid
+	delivery_valid            // we know the message is valid
+	delivery_invalid          // we know the message is invalid
+	delivery_throttled        // we can't tell if it is valid because validation throttled
+)
+
+func newPeerScore(params *PeerScoreParams) *peerScore {
+	return &peerScore{
+		params:     params,
+		peerStats:  make(map[peer.ID]*peerStats),
+		peerIPs:    make(map[string]map[peer.ID]struct{}),
+		deliveries: &messageDeliveries{records: make(map[string]*deliveryRecord)},
+		msgID:      DefaultMsgIdFn,
+	}
+}
+
+// peer score parameter validation
+func (p *PeerScoreParams) validate() error {
+	for _, params := range p.Topics {
+		err := params.validate()
+		if err != nil {
+			return err
+		}
+	}
+
+	// check that we have an app specific score; the weight can be anything (but expected positive)
+	if p.AppSpecificScore == nil {
```

Comment:
Let's also check `AppSpecificWeight` here, or maybe even remove it altogether, as this is an app-controlled function that can regulate its own score values (in contrast with the other weights that modulate pubsub behavior).

Reply:
That's a good suggestion, we can consider removing the app specific weight and just keep the score function.

Reply:
+1 to removing `AppSpecificWeight`, I don't see it adding anything.